apps/examples/src/examples/speech-bubble/SpeechBubble/helpers.tsx

```diff
@@ -1,211 +1,158 @@
+		[corners[2], corners[3]],
+		[corners[3], corners[0]],
+	]
 
-	const result = checkIntersection(handleVec, center, box)
-	if (!result) return { intersection: null, offset: null, line: null }
-	const { result: intersection, line } = result
+	let segmentsIntersection: Vec2d | null = null
+	let intersectionSegment: Vec2d[] | null = null
 
-	// lines
-	///      0
-	//  _____________
-	//  |           |
-	// 3|           | 1
-	//  |           |
-	//  -------------
-	//        2
+	// If the point inside of the box's corners?
+	const insideShape = pointInPolygon(handleInShapeSpace, corners)
 
-	const intersectionVec = new Vec2d(intersection[0].x, intersection[0].y)
-	const lineCoordinates = {
-		0: { start: new Vec2d(0, 0), end: new Vec2d(w, 0) },
-		1: { start: new Vec2d(w, 0), end: new Vec2d(w, h) },
-		2: { start: new Vec2d(0, h), end: new Vec2d(w, h) },
-		3: { start: new Vec2d(0, 0), end: new Vec2d(0, h) },
-	}
+	// We want to be sure we get an intersection, so if the point is
+	// inside the shape, push it away from the center by a big distance
+	const pointToCheck = insideShape
+		? Vec2d.Add(handleInShapeSpace, Vec2d.Sub(handleInShapeSpace, center).uni().mul(1000000))
+		: handleInShapeSpace
 
-	const { start, end } = lineCoordinates[line]
-	const whichOffset = line === 0 || line === 2 ? offset.horizontal : offset.vertical
+	// Test each segment for an intersection
+	for (const segment of segments) {
+		segmentsIntersection = intersectLineSegmentLineSegment(
+			segment[0],
+			segment[1],
+			center,
+			pointToCheck
+		)
 
-	// let's make the intersection more likely to be in the middle and also stay away from the edges
-	const adjustedIntersection = getAdjustedIntersectionPoint({
-		start,
-		end,
-		intersectionVec,
-		offset: whichOffset,
-	})
+		if (segmentsIntersection) {
+			intersectionSegment = segment
+			break
+		}
+	}
 
-	// We need the adjusted intersection to draw the tail, but the original intersection
-	// for the onBeforeUpdate handler
-	return {
-		originalIntersection: intersectionVec,
-		adjustedIntersection: adjustedIntersection,
-		offset,
-		line,
-		insideShape: result.insideShape,
+	if (!(segmentsIntersection && intersectionSegment)) {
+		throw Error("no intersection found, this shouldn't happen")
 	}
-}
 
-export const getAdjustedIntersectionPoint = ({
-	start,
-	end,
-	intersectionVec,
-	offset,
-}: {
-	start: Vec2d
-	end: Vec2d
-	intersectionVec: Vec2d
-	offset: number
-}): Vec2d | null => {
+	const [start, end] = intersectionSegment
+	const intersectionSegmentIndex = segments.indexOf(intersectionSegment)
+
 	// a normalised vector from start to end, so this can work in any direction
-	const unit = Vec2d.Sub(end, start).norm()
+	const unit = Vec2d.Sub(end, start).uni()
 
 	// Where is the intersection relative to the start?
-	const totalDistance = start.dist(end)
-	const distance = intersectionVec.dist(start)
+	const totalDistance = Vec2d.Dist(start, end)
+	const distance = Vec2d.Dist(segmentsIntersection, start)
```

Comment:
It's my fault for having both ways of doing stuff in the Vec2d library, but generally Vec2d.Dist is preferred because it works for Vec2d instances or Vec2dModel objects, save some headaches if things ever change.